### PR TITLE
Adds Create Index Building block 

### PR DESCRIPTION
**Pull request by owaiskazi19**
### Description
Adds Create Index Building block for the workflow setup and logging functionality in the project.

### Issues Resolved
Part of https://github.com/opensearch-project/opensearch-ai-flow-framework/issues/21

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -105,6 +105,7 @@ repositories {
 dependencies {
     implementation "org.opensearch:opensearch:${opensearch_version}"
     implementation 'org.junit.jupiter:junit-jupiter:5.10.0'
+    implementation "org.apache.logging.log4j:log4j-slf4j-impl:2.19.0"
```
Comment (dbwiddis) [inline build.gradle:108]: This does nothing without slf4j-api.  We already inherit log4j2 from OpenSearch so we just need to configure the logger(s).  We should use the OpenSearch config like other plugins.

File: src/main/java/org/opensearch/flowframework/FlowFrameworkPlugin.java [hunk 2]
```diff
@@ -8,11 +8,47 @@
...
+        ScriptService scriptService,
+        NamedXContentRegistry xContentRegistry,
+        Environment environment,
+        NodeEnvironment nodeEnvironment,
+        NamedWriteableRegistry namedWriteableRegistry,
+        IndexNameExpressionResolver indexNameExpressionResolver,
+        Supplier<RepositoriesService> repositoriesServiceSupplier
+    ) {
+        this.client = client;
+        CreateIndexStep createIndexStep = new CreateIndexStep(client);
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/FlowFrameworkPlugin.java:51]: Not sure why this needs to be instantiated in advance, particularly as a singleton.  As I'm iterating on draft PR #26 I think we need a generator class for these steps.

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndexStep.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+import org.opensearch.action.admin.indices.create.CreateIndexRequest;
+import org.opensearch.action.admin.indices.create.CreateIndexResponse;
+import org.opensearch.client.Client;
+import org.opensearch.common.xcontent.XContentType;
+
+import java.io.IOException;
+import java.net.URL;
+import java.util.concurrent.CompletableFuture;
+
+public class CreateIndexStep implements WorkflowStep {
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/workflow/CreateIndexStep.java:25]: See https://github.com/opensearch-project/opensearch-ai-flow-framework/pull/26/files#diff-8862e1a64f0bf9a0c3816e3b988f2faa628d1dae4e5ece3faab12a28025734b6 for ideas.

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    @Override
+    public CompletableFuture<WorkflowData> execute(List<WorkflowData> data) {
+        CompletableFuture<WorkflowData> future = new CompletableFuture<>();
+        ActionListener<CreateIndexResponse> actionListener = new ActionListener<>() {
+
+            @Override
+            public void onResponse(CreateIndexResponse createIndexResponse) {
+                logger.info("created index:{}");
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:46]:
```suggestion
                logger.info("created index: {}", createIndexResponse.index());
```
The `{}` in a log statement corresponds to an object param after it, that it evaluates the toString of.

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+                future.completeExceptionally(e);
+            }
+        };
+
+        String index = null;
+
+        for (WorkflowData workflowData : data) {
+            // Fetch index from content i.e. request body of execute API
+            Map<String, Object> content = workflowData.getContent();
+            index = (String) content.get("index");
```
Comment (dbwiddis) [inline src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:67]: As written, this will iterate all of the previous steps' data, with later ones overwriting earlier ones.  This may not be what you want.

The first element in `data` will be the initialization from the template parsing, most likely to contain any user-provided info.  The later elements are any immediate predecessor steps but ordering is not guaranteed.

TLDR: you might want a 'break;` here if index is not `null`.

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java (hunk 4 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java]: I missed that adding `index`:(

File: src/main/java/org/opensearch/flowframework/FlowFrameworkPlugin.java (hunk 2 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/FlowFrameworkPlugin.java:51]: Will update this once https://github.com/opensearch-project/opensearch-ai-flow-framework/pull/47 is in if that works?

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                break;
+            }
+        }
+
+        // TODO:
+        // 1. Create settings based on the index settings received from content
+
+        try {
+            CreateIndexRequest request = new CreateIndexRequest(index).mapping(
+                getIndexMappings("mappings/" + type + ".json"),
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:88]: Not blocking for this PR, but in my opinion index mappings should be provided by the user, since the user may want control over what fields they need. Having a static mapping is fine for now as we work towards building support for default use cases, but we'll definitely need some common mechanism to create these request bodies given the provided workflow data. This would apply to pipeline configurations and index settings as well, pretty much for any content that requires us to map fields with the user inputs.

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            Map<String, Object> content = workflowData.getContent();
+            index = (String) content.get("index-name");
+            type = (String) content.get("type");
+            if (index != null && type != null && settings != null) {
+                break;
+            }
+        }
+
+        // TODO:
+        // 1. Create settings based on the index settings received from content
```
Comment (joshpalis) [inline src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:84]: Since this initial implementation is specifically for a KNN index, we'll need at least a `knn=true` within the index settings.

File: src/test/resources/template/datademo.json [hunk 8]
```diff
@@ -3,11 +3,13 @@
         "nodes": [
             {
                 "id": "create_index",
-                "index_name": "demo"
+                "index_name": "demo",
+                "type": "knn"
```
Comment (dbwiddis) [inline src/test/resources/template/datademo.json:7]: I'm actually deleting my "datademo" json file in #47 and the "demo" JSON is in full template style.

Please write your own JSON test class if you need one, but I don't see where you use this one, so best to just undo this file change.

**Review by dbwiddis — COMMENTED** [review]
Initial review.  A few substantive and a few non-substantive requests.

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java (hunk 7 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:84]: I wouldn't say this is just for k-nn index because the mapping file is for `knn` in this PR. Based on your comment above if the mapping is provided by the user, then we do need to get the settings from the user as well or create dynamic settings based on the mappings provided for any type of index.

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+        try {
+            CreateIndexRequest request = new CreateIndexRequest(index).mapping(
+                getIndexMappings("mappings/" + type + ".json"),
+                XContentType.JSON
+            );
+            client.admin().indices().create(request, actionListener);
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:92]: Agreed, we should stop the workflow and call our exceptions here. Will iterate on this after #43 is merged.

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java (hunk 6 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:88]: I think we can have a mix here... static mappings for well-known things, user-defined mappings otherwise.  We'd just need to have some sort of syntax for the template, e.g., if user puts `{{knn}}` it will use that static mapping, or they can provide their own.

File: src/test/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStepTests.java [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+
+        inputData = new WorkflowData() {
+
+            @Override
+            public Map<String, Object> getContent() {
+                // See CreateIndexRequest ParseFields for source of content keys needed
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/test/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStepTests.java:52]: From the execute API build in #49 , all the fields will be part of the content. This comment was a leftover. Removed it.

File: src/test/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStepTests.java [hunk 11]
```diff
@@ -0,0 +1,105 @@
...
+
+        ArgumentCaptor<ActionListener> actionListenerCaptor = ArgumentCaptor.forClass(ActionListener.class);
+        CompletableFuture<WorkflowData> future = createIndexStep.execute(List.of(inputData));
+        assertFalse(future.isDone());
+        verify(indicesAdminClient, times(1)).create(any(CreateIndexRequest.class), actionListenerCaptor.capture());
+
+        actionListenerCaptor.getValue().onFailure(new Exception());
+
+        assertTrue(future.isDone());
+        assertThrows(Exception.class, () -> future.get().getContent());
```
[the remark this answers is not in the document]
Comment (owaiskazi19) [reply src/test/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStepTests.java:102]: That's a good point.

File: src/test/resources/template/datademo.json (hunk 8 above)
Comment (owaiskazi19) [reply src/test/resources/template/datademo.json]: I made this change to test locally but then I populated workflowData in test, so yes can remove these changes.

**Review by dbwiddis — APPROVED** [review]

**Review by joshpalis — APPROVED** [review]

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        String index = null;
+        String type = null;
+        Settings settings = null;
+
+        for (WorkflowData workflowData : data) {
+            Map<String, Object> content = workflowData.getContent();
+            index = (String) content.get("index-name");
+            type = (String) content.get("type");
+            if (index != null && type != null && settings != null) {
+                break;
```
Comment (jackiehanyang) [inline src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:79]: what's the point of having this break check here? We don't assign value to the `setting` variable within the loop, so it remains `null`. Isn't this means the loop will always process all items, even if it's unnecessary?

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        };
+
+        String index = null;
+        String type = null;
+        Settings settings = null;
+
+        for (WorkflowData workflowData : data) {
+            Map<String, Object> content = workflowData.getContent();
+            index = (String) content.get("index-name");
+            type = (String) content.get("type");
```
Comment (jackiehanyang) [inline src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:77]: nit: consider catching NPE when using map.get().

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java (hunk 12 above)
Comment (dbwiddis) [reply src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:79]: I interpreted this as a placeholder for future functionality, see the TODO below.

**Review by jackiehanyang — APPROVED** [review]

File: src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java (hunk 12 above)
Comment (owaiskazi19) [reply src/main/java/org/opensearch/flowframework/workflow/CreateIndex/CreateIndexStep.java:79]: Let me take this in the followup PR.
